gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,34 +312,88 @@ for _, ply in pairs(player.GetHumans()) do
     DrpCanHear[ply] = {}
 end
 
--- proxy function to take load from PlayerCanHearPlayersVoice, which is called a quadratic amount of times per tick,
--- causing a lagfest when there are many players
-local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    DrpCanHear[listener] = DrpCanHear[listener] or {}
-    if listener:IsBot() then return end
-    local shootPos = listener:GetShootPos()
-    for _, talker in ipairs(player.GetHumans()) do
-        local talkerShootPos = talker:GetShootPos()
-        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
-            (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
-                (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
-    end
-end
-hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
-    calcPlyCanHearPlayerVoice(ply)
-    if ply:IsBot() then return end
-    timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
-end)
+local gridSize = GM.Config.voiceDistance*2 -- Grid cell size is equal to the size of the diamater of player talking
+local floor = math.floor -- Caching floor as we will need to use it a lot
+
+-- Grid based position check
+timer.Create("DarkRPCanHearPlayersVoice", 1, 0, function()
```

Comment:
If this is indeed significantly more efficient in practice, we might want to lower this from 1 second to like 0.3 or something. I think this 1 second delay is actually noticeable for players. Someone racing by with their microphone on, and you either don't hear them, or for a full second.

Reply:
The time should just be a config variable.

Reply:
> If this is indeed significantly more efficient in practice, we might want to lower this from 1 second to like 0.3 or something. I think this 1 second delay is actually noticeable for players. Someone racing by with their microphone on, and you either don't hear them, or for a full second.

That is my bad, it used to be 0.5 seconds.

Reply:
> The time should just be a config variable.

Why? It's an implementation detail, this config variable would _very_ likely only confuse people. It's difficult to understand what it means, and even more difficult what the consequences will be if you set it too low or too high. This should be a value that's just good for everyone. If this code is fast enough, setting it to a low value should work for everyone, even for servers with the maximum amount of players.

Reply:
As if a server wishes to change the value because it is causing an issue or something. They have to change the core gamemode which is highly un-recommended. This leads them to if they update and forget to change the value, they may think the update of DarkRP broke something and not use it. (I have seen this happen) By making it a config and they change it, every update will use the config value causing less issues.

Reply:
The entire point is that it won't be causing issues if it's set to the right value. People will tinker with config values without the slightest clue about what they mean. Then they suddenly notice that voice doesn't work, and I get complaints about voice chat being broken. I'll then have no clue whatsoever as to why the voice would be broken, because there are no error messages.

Having a config value for this will actively cause people to break voice, and I will get the issues. It is not happening.

Reply:
Explain what the config variable does in the comment. This sort of hardcoded refresh rate is just going to lead people who DO know what they're doing to alter core files. For example, on 128 player servers, a 0.5 to even 1 second refresh rate might be preferred while at 32 players, 0.1 seconds would be. There is no reason to not provide power user features because it would "confuse" people when the simple solution is to explain what the config variable does in a comment. You could even add "-- DO NOT ALTER THIS UNLESS YOU KNOW WHAT YOU'RE DOING!"

Reply:
If the value needs to be dependent on player count, then that should be fixed by DarkRP itself, not by demanding server hosters know the intricacies of DarkRP.

Although with this PR the performance will be more dependent on how many players are near one another. If everyone is in hearing range of one another (e.g. Through noclip), you'll still get bad performance. Do you think people "who know what they're doing" should fill in this value for how often they expect many players to be near one another?

> than it is their fault for not reading.

Sure, but I'll be left not knowing it's their fault because in a year I'll have forgotten there's a config option to either completely break voice chat or slow the server to a crawl.

> lead people who DO know what they're doing to alter core files

People who know what they're doing will hopefully be people who recognise a problem in DarkRP itself. Any issue or pull request they create will be much preferred over the ones who break shit through not knowing what a config variable does, causing a problem that is extremely difficult to trace back to the source. 

The point is, no one should want to edit that value, if they do, it's either their incompetence or a bug in DarkRP. Config options are not for working around bugs.

Reply:
People have different needs. Some servers tend to have a lot of people near each other a lot (events) and thus would want a greater value. Some have more serious RP servers and thus only have a handful of people near each other at a time, thus they'd want the value as small as possible (maybe even 0) to be accurate. It's not a situation of there being one best value, but rather giving users the option for the gamemode to be versatile. Let users who know what they're doing decide whether they want accuracy or efficiency.

Reply:
With "people have different needs" one can argue for spending the rest of their lives adding config options. Why not have one for picking the algorithm itself? Maybe for really small servers the simple O(n^2) algorithm is faster. Why not have a config option for using GetShootPos or GetPos in the calculation? Why not a config option for the name of the timer? A config option for the names of the hooks? 

The line on what should be a config value and what shouldn't be  might be vague, but some things clearly should be, and others clearly shouldn't. This to me firmly falls in the latter category.

> a lot of people near each other often (events)

If darkrp is a performance bottleneck in this case, then the responsibility for that lies with DarkRP. A config option here is a statement of inability to write performance code, and a lazy excuse towards anyone who has performance problems with DarkRP. 

This config option is harmfuld for everyone. For people who don't know what they're doing it'll break the game. For the people who do know what they're doing it's a slap in the face, an insult. "Go capture your server's behavior by twiddling with a single float."

This is a PR that greatly optimises the performance of this feature. Your insistence on having a config option implies a mistrust in its performance gain. We can run some tests, I reckon 0.1 or 0.3 might be good for a 128 player server. We can then merge it and see if people start having trouble with many players near another. If it does, we can find new clever tricks like the author of this PR did here. Let's not go for a dirty cop out.

Reply:
A hardcoded number like this seems like config 101, but I give up. Let's see if the creator ever comes back to make the corrections needed.

Reply:
Sorry for the delay.

I personally stand with the config option, entirely because those are left at specific defaults pre-defined by the gamemode. We get best of both worlds as DarkRP the default setting should be a good value to be used in most cases, and those with more intricate knowledge can update the value to whatever they think is better.